graphql_compiler/schema_transformation/restrict_schema.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from copy import copy
+
+from graphql import build_ast_schema
+from graphql.language.visitor import REMOVE, Visitor, visit
+
+from ..ast_manipulation import get_ast_with_non_null_and_list_stripped
+from .utils import SchemaStructureError, get_query_type_name, get_scalar_names
+
+
+def restrict_schema(schema_ast, types_to_keep):
+    """Return new AST containing only a subset of types.
+
+    In addition to all types named in types_to_keep, the query type will also be kept. All user
+    defined scalars and enums will be kept.
+    All property fields will be kept, and only vertex fields that go to kept types will be
```

Comment:
A few more things to consider, document, and test:
- If I specify that I'd like to keep an interface, but not keep any of its implementations, is that a `SchemaStructureError`?
- If I specify that I'd like to keep an implementation of an interface but not the interface itself, is that an error, or is that okay?
- If I specify that I'd like to keep all implementations of an interface, but not the interface itself, is that an error, or is that okay?
- Do I have to specify that I'd like to keep union types, or are they kept and updated automatically? If I have to specify them, what happens if I don't?
- If I put things in `types_to_keep` that either are random strings that don't correspond to anything, or correspond to a scalar/enum/the root query type, is that an error?

There might be more considerations like the above.

Reply:
The currently version of the code works in a fairly straightforward way in my opinion, in that exactly those things you specified are kept.
So it's legal to: keep an interface and none of its implementations, keep an implementation and not its interface, keep all implementations of an interface but not the interface itself. In the latter two cases, the 'implements X' is just removed from the implementations that are kept.
The user needs to specify unions in order to keep them. If the user doesn't put the union into `types_to_keep`, then the union will be removed, along with all edges to that union.
Putting random strings or scalar/enum/query type in `types_to_keep` doesn't do anything. It's a little like how the input dictionary into `merge_schemas` works, random strings/scalars/other things that don't get used are just ignored.

Reply:
Just writing down notes from in-person discussion so I don't forget them.
- A concern was preserving an interface but not all of its implementations, since then a user could query for all instances of that interface and get back objects that are of types that don't exist in the schema.
- Another question was around `types_to_keep` containing names that don't exist in the schema, and whether that should do nothing, log to error/`warnings.warn`, or raise an error and blow up entirely.